File: erlang/euler3.erl
-module(euler3).
-export[euler3/0].

% largest prime factor of 600851475143
euler3() -> largestprimefactor(600851475143).

largestprimefactor(N) -> largestprimefactor(N, 2).

largestprimefactor(N, Candidate) when N >= Candidate * Candidate -> 
    case N rem Candidate of
    	0 -> max(Candidate, largestprimefactor(N div Candidate));
	_ -> largestprimefactor(N, Candidate + 1)
    end;
largestprimefactor(N, _) -> N.

File: clojure/euler3.clj
; Calculates the largest prime factor of a number (600851475143)

(defn largestprimefactor
    ([n] (largestprimefactor n 2))
    ([n candidate] 
        (if (< n (* candidate candidate))
	    n
            (if (zero? (mod n candidate))
	        (max candidate (largestprimefactor (quot n candidate)))
	        (largestprimefactor n (inc candidate))))))

(println 
    (largestprimefactor 600851475143))

